modules/__tests__/describeQueries.js

```diff
@@ -45,6 +45,48 @@ function describeQueries(createHistory) {
       })
     })
 
+    describe('in push', function () {
+      it('works', function (done) {
+        let steps = [
+          function (location) {
+            expect(location.pathname).toEqual('/')
+            expect(location.search).toEqual('')
+            expect(location.query).toEqual({})
+            expect(location.state).toEqual(null)
+            expect(location.action).toEqual(POP)
+
+            history.push({
+              pathname: '/home',
+              query: { the: 'query value' },
+              state: { the: 'state' }
+            })
+          },
+          function (location) {
+            expect(location.pathname).toEqual('/home')
+            expect(location.search).toEqual('?the=query+value')
+            expect(location.query).toEqual({ the: 'query value' })
+            expect(location.state).toEqual({ the: 'state' })
+            expect(location.action).toEqual(PUSH)
+
+            history.push({
+              ...location,
+              query: { other: 'query value' },
+              state: { other: 'state' }
+            })
+          },
+          function (location) {
+            expect(location.pathname).toEqual('/home')
+            expect(location.search).toEqual('?other=query+value')
+            expect(location.query).toEqual({ other: 'query value' })
```

Comment:
I would expect this object to be `{ the: 'query value', other: 'query value' }` if you want to keep with the existing behavior. I agree w you that nobody will actually want this though, and instead will just prefer to overwrite the existing query as this test appears to do. Just seems like you're introducing a breaking change here, which you said you were trying to avoid.